src/ucp/wireup/select.c

```diff
@@ -790,16 +818,46 @@ static ucs_status_t ucp_wireup_add_rndv_lane(ucp_ep_h ep,
         criteria.local_iface_flags |= UCP_WORKER_UCT_UNSIG_EVENT_CAP_FLAGS;
     }
 
-    status = ucp_wireup_select_transport(ep, address_list, address_count, &criteria,
-                                         -1, -1, 0, &rsc_index, &addr_index, &score);
-    if ((status == UCS_OK) &&
-        /* a temporary workaround to prevent the ugni uct from using rndv */
-        (strstr(ep->worker->context->tl_rscs[rsc_index].tl_rsc.tl_name, "ugni") == NULL)) {
-         ucp_wireup_add_lane_desc(lane_descs, num_lanes_p, rsc_index, addr_index,
-                                 address_list[addr_index].md_index, score,
-                                 UCP_WIREUP_LANE_USAGE_RNDV, 0);
+    if ((params->field_mask & UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE) &&
+        (params->err_mode == UCP_ERR_HANDLING_MODE_PEER)) {
+        max_lanes = ucs_min(1, ep->worker->context->config.ext.max_rndv_lanes);
+    } else {
+        max_lanes = ucs_min(UCP_MAX_RNDV_LANES, ep->worker->context->config.ext.max_rndv_lanes);
+    }
+
+    while (rndv_lanes < max_lanes) {
+        status = ucp_wireup_select_transport(ep, address_list, address_count, &criteria,
+                                             tl_bitmap, remote_md_map, 0,
+                                             &rsc_index, &addr_index, &score);
+        if (status == UCS_OK) {
+            /* Add lane description and remove all occurrences of the remote md */
+            dst_md_index = address_list[addr_index].md_index;
+            remote_md_map &= ~UCS_BIT(dst_md_index);
+            tl_bitmap = ucp_wireup_unset_tl_by_md(ep, tl_bitmap, rsc_index);
+
+            if ((strstr(ep->worker->context->tl_rscs[rsc_index].tl_rsc.tl_name, "ugni") != NULL)) {
+                /* a temporary workaround to prevent the ugni uct from using rndv */
+                continue;
+            }
+
+            rndv_lanes++;
+
+            ucp_wireup_add_lane_desc(lane_descs, num_lanes_p, rsc_index, addr_index,
+                                     dst_md_index, score,
+                                     UCP_WIREUP_LANE_USAGE_RNDV, 0);
+            if (ep->worker->context->tl_rscs[rsc_index].tl_rsc.dev_type == UCT_DEVICE_TYPE_SHM) {
+                /* In case if selected SHM transport - leave it alone, disable multirail */
+                break;
+            }
+        } else {
```

Comment:
i would just do 'if (status!=OK) break' and avoid the 'else'

Reply:
fixed